grandiso/__init__.py

```diff
@@ -298,49 +294,49 @@ def get_next_backbone_candidates(
     # in find_motifs that len(motif) == len(mapping), we will discover that the
     # mapping is "complete" even though we haven't yet checked it at all.
 
-    monomorphism_candidates = []
-
-    for mapping in tentative_results:
-        if len(mapping) == len(motif):
-            if all(
-                [
-                    host.has_edge(mapping[motif_u], mapping[motif_v])
-                    and is_edge_attr_match(
-                        (motif_u, motif_v),
-                        (mapping[motif_u], mapping[motif_v]),
-                        motif,
-                        host,
-                    )
-                    for motif_u, motif_v in motif.edges
-                ]
-            ):
-                # This is a "complete" match!
-                monomorphism_candidates.append(mapping)
-        else:
-            # This is a partial match, so we'll continue building.
-            monomorphism_candidates.append(mapping)
+    def monomorphism_candidates():
+        for mapping in tentative_results():
+            if len(mapping) == len(motif):
+                if all(
+                    [
+                        host.has_edge(mapping[motif_u], mapping[motif_v])
+                        and is_edge_attr_match(
+                            (motif_u, motif_v),
+                            (mapping[motif_u], mapping[motif_v]),
+                            motif,
+                            host,
+                        )
+                        for motif_u, motif_v in motif.edges
+                    ]
+                ):
+                    # This is a "complete" match!
+                    yield mapping
+            else:
+                # This is a partial match, so we'll continue building.
+                yield mapping
 
     if not isomorphisms_only:
-        return monomorphism_candidates
+        yield from monomorphism_candidates()
+        return
 
     # Additionally, if isomorphisms_only == True, we can use this opportunity
     # to confirm that no spurious edges exist in the induced subgraph:
-    isomorphism_candidates = []
-    for result in monomorphism_candidates:
-        for (motif_u, motif_v) in itertools.product(result.keys(), result.keys()):
-            # if the motif has this edge, then it doesn't rule any of the
-            # above results out as an isomorphism.
-            # if the motif does NOT have the edge, then NO RESULT may have
-            # the equivalent edge in the host graph:
-            if not motif.has_edge(motif_u, motif_v) and host.has_edge(
-                result[motif_u], result[motif_v]
-            ):
-                # this is a violation.
-                break
-        else:
-            isomorphism_candidates.append(result)
-    return isomorphism_candidates
+    def isomorphism_candidates():
```

Comment:
sorry for the quiet; doing a little benchmarking right now on the performance implications of calling a nested func vs unwrapping these!

Reply:
Sounds good. I tried to make the change as minimally invasive as possible but I agree that having them as separate functions would be cleaner.

Let me know if you come across anything that seems slower. In the worst case scenario, it goes through all the data and discovers there are no matches (i.e. a full scan). 

I also think the grand project is a pretty cool idea, putting a networkx interface on different backends such as a SQL database. I could see where indexes could make full scan operations much faster.

Reply:
@davidmezzetti I got about +10μs benchmarking locally with ASV, which I think is fair to say is in the noise :) nice work!

I'd LOVE to have better indexing in Grand; I've been aiming for feature-completeness first and then adding in "shortcuts" and optimizations; would love to chat about some gameplans if you're interested!

Reply:
Sounds good to me. I think having a networkx dialect applied towards something like a SQLite backend with indexes could be a powerful in-memory graph solution.

Things like an attribute search could return immediately regardless if there was/wasn't data for example (i.e. no full scans). In the case of SQLite, you'd get the native code and smart memory management for free. But still have the ease-of-use of networkx.

Reply:
COMPLETELY agree, would love to have a sqlite true high speed implementation!